src/staking/block_validator.cpp

```diff
@@ -258,6 +319,50 @@ class BlockValidatorImpl : public AbstractBlockValidator {
     return CheckCoinbaseTransactionInternal(block, coinbase_tx, nullptr, nullptr);
   }
 
+  BlockValidationResult CheckTransaction(
+      const CTransaction &tx) const override {
+    BlockValidationResult result;
+    if (tx.vin.empty()) {
+      return BlockValidationResult(Error::INVALID_TRANSACTION_NO_INPUTS);
+    }
+    if (tx.vout.empty()) {
+      return BlockValidationResult(Error::INVALID_TRANSACTION_NO_OUTPUTS);
+    }
+    const std::size_t size = GetSerializeSize(tx, SER_NETWORK, PROTOCOL_VERSION | SERIALIZE_TRANSACTION_NO_WITNESS);
+    const std::size_t weight = size * m_blockchain_behavior->GetParameters().witness_scale_factor;
+    if (weight > m_blockchain_behavior->GetParameters().maximum_block_weight) {
+      return BlockValidationResult(Error::INVALID_TRANSACTION_TOO_BIG);
+    }
+    std::set<COutPoint> outpoints;
+    for (const CTxIn &txin : tx.vin) {
+      if (!outpoints.insert(txin.prevout).second) {
+        return BlockValidationResult(Error::INVALID_TRANSACTION_DUPLICATE_INPUTS);
+      }
+      if (txin.prevout.IsNull()) {
+        if (tx.IsCoinBase()) {
```

Comment:
Shouldn't the `prevout` be the output of the staking tx?

Reply:
Yes, and this is checked in the `StakeValidator` and in `CheckInputs` which is invoked on `ConnectBlock`.

The `CheckTransaction` checks are sort-of "cosmetic", they only check that a transaction "looks" right, i.e. is well-formed.

A null-input is only allowed for the meta-input in the coinbase transaction.

Thank you for pointing this out, the check should not just check whether we're in a coinbase transaction but also whether it's the zeroth input.

Reply:
yeah I was guessing so too.

Reply:
Fixed in https://github.com/dtr-org/unit-e/pull/1106/commits/6d34798b21f29bf02eb0f4c7c82a0676aa21a13e